File: lib/wType/event.cpp
#include "event.h"

W::Event::Event():
    Object(),
    system(false),
    destination(),
    sender(0),
    data(0)
{

}

W::Event::Event(const W::Address& p_addr, const W::Object& p_data, bool p_system):
    Object(),
    system(p_system),
    destination(p_addr),
    sender(0),
    data(p_data.copy())
{

}

W::Event::Event(const W::Address& p_addr, W::Object* p_data, bool p_system):
    Object(),
    system(p_system),
    destination(p_addr),
    sender(0),
    data(p_data)
{

}


W::Event::Event(const W::Event& original):
    Object(),
    system(original.system),
    destination(original.destination),
    sender(original.sender),
    data(original.data->copy())
{

}

W::Event::~Event()
{
    delete data;
}

W::Event& W::Event::operator=(const W::Event& original)
{
    if (&original != this)
    {
        delete data;
        
        system = original.system;
        destination = original.destination;
        sender = original.sender;
        data = original.data->copy();
    }
    return *this;
}

W::Object* W::Event::copy() const
{
    return new Event(*this);
}

W::Object::objectType W::Event::getType() const
{
    return Event::type;
}

W::Object::size_type W::Event::length() const
{
    size_type my_size = 2;
    my_size += destination.length();
    my_size += data->length();
    return my_size;
}

W::Object::StdStr W::Event::toString() const
{
    StdStr result;
    
    result += "{";
    
    result += "system: ";
    result += system.toString();
    
    result += ", desitnation: ";
    result += destination.toString();
    
    result += ", sender: ";
    result += sender.toString();
    
    result += ", data: ";
    result += data->toString();
    
    result += "}";
    
    return result;
}

void W::Event::serialize(W::ByteArray& out) const
{
    system.serialize(out);
    if (!system)
    {
        destination.serialize(out);
        sender.serialize(out);
    }
    
    out.push8(data->getType());
    data->serialize(out);
}

void W::Event::deserialize(W::ByteArray& in)
{
    system.deserialize(in);
    if (!system)
    {
        destination.deserialize(in);
        sender.deserialize(in);
    }
    
    delete data;
    
    data = Object::fromByteArray(in);
}

bool W::Event::isSystem() const
{
    return system;
}

const W::Address& W::Event::getDestination() const
{
    return destination;
}

uint64_t W::Event::getSenderId() const
{
    return sender;
}

const W::Object& W::Event::getData() const
{
    return *data;
}

void W::Event::setSenderId(const W::Uint64& senderId)
{
    sender = senderId;
}

void W::Event::setSenderId(uint64_t senderId)
{
    sender = Uint64(senderId);
}

bool W::Event::operator==(const W::Object& other) const
{
    if (sameType(other)) {
        const W::Event& oev = static_cast<const W::Event&>(other);
        return destination == oev.destination && system == oev.system && sender == oev.sender && *data == *(oev.data);
    } else {
        return false;
    }
}

W::Object::size_type W::Event::size() const
{
    size_type sz = system.size() + data->size() + 1;
    if (!system)
    {
        sz += destination.size() + sender.size();
    }
    return sz;
}

File: libjs/wType/event.js
"use strict";
var Object = require("./object");
var Uint64 = require("./uint64");
var Address = require("./address");
var Boolean = require("./boolean");

var Event = Object.inherit({
    "className": "Event",
    "constructor": function(addr, object, isSystem) {
        Object.fn.constructor.call(this);
        
        if (!(object instanceof Object) && object !== undefined) {
            throw new Error("Wrong arguments to construct Event");
        }
        if (!(addr instanceof Address) && addr !== undefined) {
            throw new Error("Wrong arguments to construct Event");
        }
        
        this._destination = addr !== undefined ? addr : new Address();
        this._data = object !== undefined ? object : new Object();
        this._senderId = new Uint64();
        this._system = new Boolean(isSystem);
    },
    "destructor": function() {
        this.clear();
        
        Object.fn.destructor.call(this);
    },
    "==": function(other) {
        if (this.getType() !== other.getType()) {
            return false;
        }
        return  this._destination["=="](other._destination) &&
                this._system["=="](other._system) &&
                this._senderId["=="](other._senderId) &&
                this._data["=="](other._data)
    },
    "clear": function() {
        this._system.destructor();
        this._destination.destructor();
        this._senderId.destructor();
        this._data.destructor();
    },
    "clone": function() {
        var clone = new Event();
        
        clone._destination = this._destination.clone();
        clone._data = this._data.clone();
        clone._senderId = this._senderId.clone();
        clone._system = this._system.clone();
        
        return clone;
    },
    "deserialize": function(ba) {
        this._system = new Boolean();
        this._system.deserialize(ba);
        
        if (!this.isSystem()) { 
            this._destination = new Address();
            this._destination.deserialize(ba);
        
            this._senderId = new Uint64();
            this._senderId.deserialize(ba);
        }
        
        this._data = Object.fromByteArray(ba);
    },
    "getData": function() {
        return this._data;
    },
    "getDestination": function() {
        return this._destination;
    },
    "getSenderId": function() {
        return this._senderId;
    },
    "isSystem": function() {
        return this._system.valueOf();
    },
    "length": function() {
        return 2 + this._destination.length() + this._data.length();
    },
    "serialize": function(ba) {
        this._system.serialize(ba);
        
        if (!this.isSystem()) {
            this._destination.serialize(ba);
            this._senderId.serialize(ba);
        }
        
        ba.push8(this._data.getType());
        this._data.serialize(ba);
    },
    "setSenderId": function(id) {
        if (!(id instanceof Uint64)) {
            throw new Error("Can't set id, which is not Uint64");
        }
        this._senderId = id;
    },
    "size": function() {
        var size = this._system.size() + this._data.size() + 1
        if (!this.isSystem()) {
            size += this._senderId.size() + this._destination.size();
        }
        return size;
    },
    "toString": function() {
        var str = "{";
        
        str += "system: " + this._system.toString();
        str += " destination: " + this._destination.toString();
        str += " sender: " + this._senderId.toString();
        str += " data: " + this._data.toString();
        
        str += "}";
        
        return str;
    }
});

module.exports = Event;
